File: src/main/java/de/bonndan/nivio/output/map/NPath.java
package de.bonndan.nivio.output.map;

import j2html.tags.ContainerTag;
import j2html.tags.DomContent;

import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.List;

import static de.bonndan.nivio.output.map.SvgTagCreator.g;

class NPath extends Component {

    public static final String MARKER = "▸";
    private final TilePath tilePath;
    private final String fill;
    private final ItemMapItem.Relation relation;

    NPath(TilePath tilePath, String fill, ItemMapItem.Relation relation) {
        this.tilePath = tilePath;
        this.fill = fill;
        this.relation = relation;
    }

    public DomContent render() {

        var fillId = (fill) != null ? "#" + fill : null;
        var stringPath = tilePath.getPoints();

        BezierPath bezierPath = new BezierPath();
        bezierPath.parsePathString(stringPath);

        if ("PROVIDER".equals(relation.type)) {
            return g(
                    SvgTagCreator.path()
                            .attr("d", stringPath)
                            .attr("stroke", fillId)
                            .attr("data-target", relation.target)
                            .attr("data-type", relation.type)
                    ,
                    label(bezierPath, fillId)
            );
        }

        List<ContainerTag> markers = new ArrayList<>();
        float pieces = bezierPath.path.curveLength / 20;
        float pct = 100 / pieces;
        for (float i = 0; i < 1; i += pct / 100) {
            Point2D.Float point1 = bezierPath.eval(i);
            Point2D.Float point2 = bezierPath.eval(i + 0.01f);
            markers.add(this.marker(point1, point2, fillId));
        }

        return g(markers.toArray(DomContent[]::new));
    }

    private ContainerTag marker(Point2D.Float point, Point2D.Float point2, String fillId) {
        return alongPath(MARKER, point, point2, fillId, -10);
    }

    private ContainerTag label(BezierPath bezierPath, String fillId) {
        Point2D.Float point = bezierPath.eval(0.5f);
        Point2D.Float point2 = bezierPath.eval(0.51f);
        return alongPath(relation.format, point, point2, fillId, 10);
    }

    private ContainerTag alongPath(String text, Point2D.Float point, Point2D.Float point2, String fillId, int xOffset) {

        var degrees = Math.atan2((point2.y - point.y), (point2.x - point.x)) * 180 / Math.PI;
        String transform = "translate(" + Math.floor(point.getX()) + ' ' + point.getY() + ") rotate(" + degrees + " 0 0)";

        if (text == null) {
            text = "";
        }
        return SvgTagCreator.text(text)
                .attr("x", xOffset)
                .attr("y", 0)
                .attr("fill", fillId)
                .attr("width", 10)
                .attr("height", 10)
                .attr("transform", transform);
    }
}

File: src/main/app/src/hexmap/NPath.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { svgPathProperties } from 'svg-path-properties';

class NPath extends Component {
  static propTypes = {
    fill: PropTypes.string,
    tilePath: PropTypes.object,
    layout: PropTypes.object,
    realtion: PropTypes.object,
  };
  static contextTypes = {
    layout: PropTypes.object,
  };

  getPoints() {
    const { tilePath } = this.props;
    const { layout } = this.context;
    return tilePath.getPoints(layout);
  }

  render() {
    const { fill, relation } = this.props;
    const fillId = fill ? `#${fill}` : null;
    let path = this.getPoints();

    if (relation.type === 'PROVIDER') {
      return <path d={path} stroke={fillId} />;
    }

    const properties = new svgPathProperties(path);
    const separation = 15;
    let markers = [];
    for (var j = 0; j < Math.floor(properties.getTotalLength() / separation); j++) {
      var pos = properties.getPropertiesAtLength(separation * j);
      markers.push(this.marker(pos, fillId));
    }

    return markers;
  }

  marker(pos, fillId) {
    let degrees = (Math.atan(pos.tangentY / pos.tangentX) * 180) / Math.PI;
    if (pos.tangentX < 0 && pos.tangentY > 0)
      //TODO check
      degrees += 180;

    let transform =
      'translate(' + Math.floor(pos.x) + ' ' + pos.y + ') rotate(' + degrees + ' 0 0)';
    return (
      <text x='-10' y='0' fill={fillId} width='10' height='10' transform={transform}>
        &#10148;
      </text>
    );
  }
}

export default NPath;
